webui/react/src/services/stream/stream.ts

```diff
@@ -0,0 +1,151 @@
+                        const stream_key = trimEnd(k, '_deleted') as Streamable;
+                        const deleted_keys = decode_keys(val as string);
+                        this.#curEntity?.[stream_key].keyCache.delete_msg(deleted_keys);
+                        onDelete(stream_key, deleted_keys);
+                    } else {
+                        this.#curEntity?.[StreamEntityMap[k]].keyCache.upsert([val.id], val.seq);
+                        onUpsert(msg);
+                    }
+                });
+
+            }
+
+        };
+    }
+
+    async retry(): Promise<boolean> {
+        try {
+            const backoff = backoffs[this.#retries];
+            this.#retries += 1;
+            await sleep(backoff);
+            return true;
+        } catch {
+            return false;
+        }
+    }
+
+    #sameAsCur(entity: Record<Streamable, Entity>): boolean {
+        if (!this.#curEntity) return false;
+        let same = true;
+        forEach(entity, (val, k) => {
+            if (!this.#curEntity?.[k as Streamable].spec.equals(val.spec)) same = false;
+        });
+        return same;
+    }
+
+    #sendSpec(entity?: Record<Streamable, Entity>): void {
+        if (!entity || this.#sameAsCur(entity)) return;
+
+        this.#numSyncs += 1;
+        const sync_id = this.#numSyncs.toString();
+        const spec = reduce(entity, (spec, ent, k) => {
+            spec['known'][k] = ent.keyCache.known();
+            spec['subscribe'][k] = {
+                ...ent.spec.toWire(),
+                since: ent.keyCache.maxSeq(),
+            };
+            return spec;
+        }, { known: {}, subscribe: {}, sync_id: sync_id } as Record<string, any>);
+
+        this.#ws.send(JSON.stringify(spec));
+        this.#curEntity = entity;
+        this.#syncSent = sync_id;
+    }
+
+    advanceSubscription(): void {
+        if (this.#ws.readyState !== this.#ws.OPEN) return;
+        if (!this.#curEntity && !this.#entities) return;
+
+        let spec: Record<Streamable, Entity> | undefined;
+        if (!this.#syncSent) {
+            if (this.#curEntity) {
+                spec = this.#curEntity;
+            } else {
+                spec = this.#entities.shift();
+            }
+            this.#sendSpec(spec);
+        }
+        if (this.#entities && this.#syncComplete === this.#syncSent) {
+            spec = this.#entities.shift();
+            this.#sendSpec(spec);
+        }
+    }
+
+    public subscribe(spec: StreamSpec, known?: Array<number>): Stream {
+        const curSpec = this.#curEntity?.[spec.id()];
+        const keyCache = curSpec && !curSpec.spec.equals(spec) ? curSpec.keyCache : new KeyCache();
+        if (known) {
+            keyCache.upsert(known);
+        }
+
+        this.#entities.push({ ...this.#curEntity, [spec.id()]: { keyCache, spec } });
+        this.advanceSubscription();
+
+        return this;
+    }
+
+    public close(): void {
+        this.#ws.close();
```

Comment:
maybe a dumb question... but does `#ws.close()` trigger a `#ws.onerror()` callback?  I can't remember...

Reply:
Also, do you need to save the fact that you closed down, so that if you are in a retry backoff, you don't trigger the open event after the retry is completed?